core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,492 @@
+/*
+ * Copyright 2023 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.CONNECTING;
+import static io.grpc.ConnectivityState.IDLE;
+import static io.grpc.ConnectivityState.READY;
+import static io.grpc.ConnectivityState.SHUTDOWN;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.Lists;
+import io.grpc.Attributes;
+import io.grpc.ConnectivityState;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.ExperimentalApi;
+import io.grpc.LoadBalancer;
+import io.grpc.Status;
+import java.net.SocketAddress;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link LoadBalancer} that provides no load-balancing over the addresses from the {@link
+ * io.grpc.NameResolver}.  The channel's default behavior is used, which is walking down the address
+ * list and sticking to the first that works.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/10383")
+final class PickFirstLeafLoadBalancer extends LoadBalancer {
+  private final Helper helper;
+  private final Map<SocketAddress, Subchannel> subchannels = new HashMap<>();
+  private final Map<Subchannel, ConnectivityState> states = new HashMap<>();
```

Comment:
You don't even need to wrap the Subchannel here. Just have a class to hold/manage the subchannel. You can put more logic in it if you want, but at its most basic, it is just a data holder.

```java
private final Map<SocketAddress, SubchannelData> subchannels = new HashMap<>();

static final class SubchannelData {
  final Subchannel subchannel;
  ConnectivityState state;
  boolean deleted;
}

...
    SubchannelData data = new SubchannelData(subchannel, IDLE);
    subchannels.put(addr, data);
    subchannel.start(new SubchannelStateListener() {
      @Override
      public void onSubchannelState(ConnectivityStateInfo stateInfo) {
        processSubchannelState(data, stateInfo);
      }
    });
```

Reply:
I tried implementing this and it seems like there may be some tradeoffs. While it simplifies the logic in some ways, it complicates it in others by requiring us to check for nulls much more frequently. I'm leaning towards advocating for the current state and subchannel map, but what do you think?

Reply:
When does it require additional null checks? We can guarantee/checkNotNull() that `data.subchannel` is not null. So that means only `data` could be null, which should be the same as checking that subchannel is null.

Reply:
Before, we could just compare subchannels via `subchannelA == subchannelB`, which would return false if either was null. Having to check to see if `subchannelData` is null prior to using the subchannel adds pretty much the same amount of lines that we shave off from having this data structure.

Reply:
Where would we do that comparison? Seems we could compare `dataA == dataB` instead of `dataA.subchannel == dataB.subchannel`.

I actually don't see any places we do `subchannels.get()` where it could return `null`. (But that's partly because there is already protection using `subchannels.containsKey()`.)

Reply:
Ah you're right. I changed L166 from something like `subchannels.get(currentAddress).equals(subchannel)` to  `subchannels.containsKey()` a bit ago and it seems like this fixes the issue. I'll push with the new class.